SystemHelper.kt:1:1: Internal Error (no-unused-imports)

## Expected Behavior
I am not given this warning.

## Observed Behavior
I am given this warning.
```
bot-kt % ktlint -F "src/main/kotlin/org/kamiblue/botkt/"
/home/mika/projects/bot-kt/src/main/kotlin/org/kamiblue/botkt/helpers/SystemHelper.kt:1:1: Internal Error (no-unused-imports). Please create a ticket at https://github.com/pinterest/ktlint/issues (if possible, provide the source code that triggered an error) (cannot be auto-corrected)
```

## Steps to Reproduce
My [.editorconfig](https://github.com/kami-blue/bot-kt/blob/dad84e0f74618fed68ce89a5e77f15ddff4ea8f3/.editorconfig)
Affected file: [SystemHelper.kt](https://github.com/kami-blue/bot-kt/blob/dad84e0f74618fed68ce89a5e77f15ddff4ea8f3/src/main/kotlin/org/kamiblue/botkt/helpers/SystemHelper.kt#L1)
Run `ktlint -F "src/main/kotlin/org/kamiblue/botkt/"`

## Your Environment
* Version of ktlint used: `0.39.0`
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): `ktlint -F "src/main/kotlin/org/kamiblue/botkt/"`
* Version of Gradle used (if applicable): `6.7.1`
* Operating System and version: `Linux strawberry 5.9.11-arch2-1 #1 SMP PREEMPT Sat, 28 Nov 2020 02:07:22 +0000 x86_64 GNU/Linux`
* Link to your project (if it's a public repository): [bot-kt](https://github.com/kami-blue/bot-kt/), building instructions in [README.md](https://github.com/kami-blue/bot-kt#contributing)

Error stacktrace:
```java
[DEBUG] Internal Error (no-unused-imports)
com.pinterest.ktlint.core.RuleExecutionException: java.lang.IllegalStateException: No operation node for DOT_QUALIFIED_EXPRESSION. Children: [CALL_EXPRESSION]
        at com.pinterest.ktlint.core.KtLint$format$2.invoke(KtLint.kt:346)
        at com.pinterest.ktlint.core.KtLint$format$2.invoke(KtLint.kt:27)
        at com.pinterest.ktlint.core.KtLint$visitor$2$1.invoke(KtLint.kt:241)
        at com.pinterest.ktlint.core.KtLint$visitor$2$1.invoke(KtLint.kt:27)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:225)
        at com.pinterest.ktlint.core.KtLint$visitor$2.invoke(KtLint.kt:239)
        at com.pinterest.ktlint.core.KtLint$visitor$2.invoke(KtLint.kt:27)
        at com.pinterest.ktlint.core.KtLint.format(KtLint.kt:326)
        at com.pinterest.ktlint.internal.FileUtilsKt.formatFile(FileUtils.kt:99)
        at com.pinterest.ktlint.KtlintCommandLine.process(Main.kt:358)
        at com.pinterest.ktlint.KtlintCommandLine.access$process(Main.kt:124)
        at com.pinterest.ktlint.KtlintCommandLine$lintFiles$2$1.call(Main.kt:281)
        at com.pinterest.ktlint.KtlintCommandLine$lintFiles$2$1.call(Main.kt:124)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: No operation node for DOT_QUALIFIED_EXPRESSION. Children: [CALL_EXPRESSION]
        at org.jetbrains.kotlin.psi.KtQualifiedExpression$DefaultImpls.getOperationTokenNode(KtQualifiedExpression.kt:35)
        at org.jetbrains.kotlin.psi.KtDotQualifiedExpression.getOperationTokenNode(KtDotQualifiedExpression.kt:26)
        at org.jetbrains.kotlin.psi.KtQualifiedExpression$DefaultImpls.getExpression(KtQualifiedExpression.kt:43)
        at org.jetbrains.kotlin.psi.KtQualifiedExpression$DefaultImpls.getSelectorExpression(KtQualifiedExpression.kt:32)
        at org.jetbrains.kotlin.psi.KtDotQualifiedExpression.getSelectorExpression(KtDotQualifiedExpression.kt:56)
        at com.pinterest.ktlint.ruleset.standard.NoUnusedImportsRule.parentDotQualifiedExpression(NoUnusedImportsRule.kt:177)
        at com.pinterest.ktlint.ruleset.standard.NoUnusedImportsRule.access$parentDotQualifiedExpression(NoUnusedImportsRule.kt:23)
        at com.pinterest.ktlint.ruleset.standard.NoUnusedImportsRule$visit$1.invoke(NoUnusedImportsRule.kt:100)
        at com.pinterest.ktlint.ruleset.standard.NoUnusedImportsRule$visit$1.invoke(NoUnusedImportsRule.kt:23)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:225)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.core.ast.PackageKt.visit(package.kt:226)
        at com.pinterest.ktlint.ruleset.standard.NoUnusedImportsRule.visit(NoUnusedImportsRule.kt:87)
        at com.pinterest.ktlint.core.KtLint$format$2.invoke(KtLint.kt:333)
        ... 16 more
/tmp/SystemHelper.kt:1:1: Internal Error (no-unused-imports). Please create a ticket at https://github.com/pinterest/ktlint/issues (if possible, provide the source code that triggered an error) (cannot be auto-corrected)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - Fix false positive when argument list is after multiline dot-qualified expression (`argument-list-wrapping`) ([#893](https://github.com/pinterest/ktlint/issues/893))
 - Fix indentation for function types after a newline (`indent`) ([#918](https://github.com/pinterest/ktlint/issues/918))
 - Don't remove the equals sign for a default argument (`no-line-break-before-assignment`) ([#1039](https://github.com/pinterest/ktlint/issues/1039))
+- Fix internal error in `no-unused-imports` ([#1040](https://github.com/pinterest/ktlint/issues/1040))
 
 ### Changed
 - Update Gradle shadow plugin to `6.1.0` version

--- a/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/StringTemplateRule.kt
+++ b/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/StringTemplateRule.kt
@@ -2,19 +2,18 @@ package com.pinterest.ktlint.ruleset.standard
 
 import com.pinterest.ktlint.core.Rule
 import com.pinterest.ktlint.core.ast.ElementType.CLOSING_QUOTE
-import com.pinterest.ktlint.core.ast.ElementType.DOT
 import com.pinterest.ktlint.core.ast.ElementType.DOT_QUALIFIED_EXPRESSION
 import com.pinterest.ktlint.core.ast.ElementType.LITERAL_STRING_TEMPLATE_ENTRY
 import com.pinterest.ktlint.core.ast.ElementType.LONG_STRING_TEMPLATE_ENTRY
 import com.pinterest.ktlint.core.ast.ElementType.LONG_TEMPLATE_ENTRY_END
 import com.pinterest.ktlint.core.ast.ElementType.LONG_TEMPLATE_ENTRY_START
 import com.pinterest.ktlint.core.ast.ElementType.REGULAR_STRING_PART
-import com.pinterest.ktlint.core.ast.ElementType.SUPER_EXPRESSION
 import org.jetbrains.kotlin.com.intellij.lang.ASTNode
 import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LeafPsiElement
 import org.jetbrains.kotlin.psi.KtBlockStringTemplateEntry
 import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
 import org.jetbrains.kotlin.psi.KtNameReferenceExpression
+import org.jetbrains.kotlin.psi.KtSuperExpression
 import org.jetbrains.kotlin.psi.KtThisExpression
 
 class StringTemplateRule : Rule("string-template") {
@@ -38,20 +37,17 @@ class StringTemplateRule : Rule("string-template") {
 */
         if (elementType == LONG_STRING_TEMPLATE_ENTRY) {
             var entryExpression = (node.psi as? KtBlockStringTemplateEntry)?.expression
-            val entryStart = node.firstChildNode
-            val dotQualifiedExpression = entryStart.treeNext
-            if (dotQualifiedExpression?.elementType == DOT_QUALIFIED_EXPRESSION) {
-                val callExpression = dotQualifiedExpression.lastChildNode
-                val dot = callExpression.treePrev
-                if (dot?.elementType == DOT &&
-                    callExpression.text == "toString()" &&
-                    dotQualifiedExpression.firstChildNode?.elementType != SUPER_EXPRESSION
-                ) {
-                    emit(dot.startOffset, "Redundant \"toString()\" call in string template", true)
+            if (entryExpression is KtDotQualifiedExpression) {
+                val receiver = entryExpression.receiverExpression
+                if (entryExpression.selectorExpression?.text == "toString()" && receiver !is KtSuperExpression) {
+                    emit(
+                        entryExpression.operationTokenNode.startOffset,
+                        "Redundant \"toString()\" call in string template",
+                        true
+                    )
                     if (autoCorrect) {
-                        entryExpression = (entryExpression as? KtDotQualifiedExpression)?.receiverExpression
-                        node.removeChild(dot)
-                        node.removeChild(callExpression)
+                        entryExpression.replace(receiver)
+                        entryExpression = receiver
                     }
                 }
             }

--- a/ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/StandardRuleSetIntegrationTest.kt
+++ b/ktlint-ruleset-standard/src/test/kotlin/com/pinterest/ktlint/ruleset/standard/StandardRuleSetIntegrationTest.kt
@@ -1,0 +1,33 @@
+package com.pinterest.ktlint.ruleset.standard
+
+import com.pinterest.ktlint.core.KtLint
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Test
+
+class StandardRuleSetIntegrationTest {
+
+    @Test
+    fun `test string-template wit unused-imports integration`() {
+        val actual = KtLint.format(
+            KtLint.Params(
+                text = """
+                fun foo() = 1
+                fun test() {
+                    println("${'$'}{foo().toString()}")
+                }
+                """.trimIndent(),
+                ruleSets = listOf(StandardRuleSetProvider().get()),
+                cb = { _, _ -> }
+            )
+        )
+        assertThat(actual).isEqualTo(
+            """
+            fun foo() = 1
+            fun test() {
+                println("${'$'}{foo()}")
+            }
+
+            """.trimIndent()
+        )
+    }
+}

